main.cc

```diff
@@ -1277,6 +1277,9 @@ int main(int ac, char** av) {
             db::get_batchlog_manager().invoke_on_all([] (db::batchlog_manager& b) {
                 return b.start();
             }).get();
+            auto stop_batchlog_manager = defer_verbose_shutdown("batchlog manager", [] {
+                db::get_batchlog_manager().invoke_on_all(&db::batchlog_manager::stop).get();
+            });
```

Comment:
stop() is not idempotent due to the std::exchange(). The first stop() will wait, the second stop() will not.

If you want to make stop() idempotent, use a shared_future. The best is not to rely on stop() idempotency, it's just tricky. Call stop() exactly one time, and abort() to abort the loop.

Reply:
okay, makes sense

Reply:
fixed in v4